File: app/src/main/java/me/lengyu/qedge/utils/ClassUtils.java
package me.lengyu.qedge.utils;

public class ClassUtils {

    public static ClassLoader hostClassLoader;

    static {
        try {
            Class<?> activityThreadClass = Class.forName("android.app.ActivityThread");
            Object currentActivityThread = activityThreadClass.getMethod("currentActivityThread").invoke(null);
            if (currentActivityThread != null) {
                java.lang.reflect.Field mPackagesField = activityThreadClass.getDeclaredField("mPackages");
                mPackagesField.setAccessible(true);
                Object mPackages = mPackagesField.get(currentActivityThread);
                if (mPackages instanceof android.util.ArrayMap) {
                    android.util.ArrayMap<?, ?> packages = (android.util.ArrayMap<?, ?>) mPackages;
                    if (!packages.isEmpty()) {
                        Object packageInfo = packages.valueAt(0);
                        Class<?> loadedApkClass = Class.forName("android.app.LoadedApk");
                        java.lang.reflect.Field mClassLoaderField = loadedApkClass.getDeclaredField("mClassLoader");
                        mClassLoaderField.setAccessible(true);
                        hostClassLoader = (ClassLoader) mClassLoaderField.get(packageInfo);
                    }
                }
            }
        } catch (Throwable e) {
            e.printStackTrace();
        }
        if (hostClassLoader == null) {
            hostClassLoader = ClassUtils.class.getClassLoader();
        }
    }

    public static ClassLoader getHostClassLoader() {
        return hostClassLoader;
    }
}

File: app/src/main/java/me/lengyu/qedge/utils/reflect/ClassUtils.kt
package me.lengyu.qedge.utils.reflect

import me.lengyu.qedge.utils.HostInfo

object ClassUtils {
    private var _hostClassLoader: ClassLoader? = null

    var hostClassLoader: ClassLoader
        get() = _hostClassLoader ?: throw IllegalStateException("HostClassLoader not initialized")
        set(value) {
            _hostClassLoader = value
        }

    val moduleClassLoader: ClassLoader by lazy { this::class.java.classLoader!! }

    fun loadClassOrNull(className: String): Class<*>? {
        return runCatching { hostClassLoader.loadClass(className) }.getOrNull()
    }

    fun loadClass(className: String): Class<*> {
        return hostClassLoader.loadClass(className)
    }
}

val String.clazz: Class<*>?
    get() = ClassUtils.loadClassOrNull(this)

val String.toClass: Class<*>
    get() = ClassUtils.loadClass(this)